src/crimson/os/seastore/segment_manager/zns.cc

```diff
@@ -603,4 +608,72 @@ Segment::write_ertr::future<> ZNSSegment::write(
   return manager.segment_write(paddr_t::make_seg_paddr(id, offset), bl);
 }
 
+Segment::write_ertr::future<> ZNSSegment::write_padding_bytes(
+  size_t padding_bytes, seastore_off_t tail_offset)
+{
+  LOG_PREFIX(ZNSSegment::write_padding_bytes);
+  DEBUG("Writing {} padding bytes to segment {} at wp {}",
+                                                padding_bytes, id, write_pointer);
+
+  return crimson::repeat([FNAME, padding_bytes, tail_offset, this] () mutable {
+    size_t bufsize = 0;
+    if (padding_bytes >= MAX_WRITE_BUF_SIZE) {
+      bufsize = MAX_WRITE_BUF_SIZE;
+    } else {
+      bufsize = MIN_WRITE_BUF_SIZE;
+    }
+
+    padding_bytes -= bufsize;
+    bufferptr bp(ceph::buffer::create_page_aligned(bufsize));
+    bp.zero();
+    bufferlist padd_bl;
+    padd_bl.append(bp);
+    return write(write_pointer, padd_bl).safe_then([FNAME, tail_offset, this]() {
+      if (write_pointer == tail_offset) {
```

Comment:
I think the the name tail_offset is misleading. It is the offset till the wp has to be advanced.
I willl change the name to something more appropriate(like target_offset).

My thought process was that this approach is more modular and it can be reused by other callers if needed. advance_wp() will become a big function, so I split it to 2 functions. If you think they both needs to be merged, I will merge them both.

Reply:
> padding_bytes can be used here to stop the iteration instead of the another tail_offset.

I meant the offset is not needed in this function, the following logic would be sufficient:
```c++
return write(write_pointer, padd_bl
).safe_then([&padding_bytes]() {
  if (padding_bytes == 0) {
    return write_ertr::make_ready_future<seastar::stop_iteration>(seastar::stop_iteration::yes);
  } else {
    return write_ertr::make_ready_future<seastar::stop_iteration>(seastar::stop_iteration::no);
  }
...
```

> My thought process was that this approach is more modular and it can be reused by other callers if needed. advance_wp() will become a big function, so I split it to 2 functions. If you think they both needs to be merged, I will merge them both.

I'm fine with either, if choose to split, please move `write_padding_bytes()` to the private section.

Reply:
Ok, will move to private section.
Will fix the other part. Though I think I saw some issue like the padding_bytes value was not retained across the loops or something like that. Will verify.

Reply:
> Though I think I saw some issue like the padding_bytes value was not retained across the loops or something like that.

If it's true, use `do_with()` to extend the life of `padding_bytes` inside the loop, that should be always working.

Reply:
I just verified that your earlier suggestion works. Changed it.